src/jit/decomposelongs.cpp

```diff
@@ -850,6 +850,90 @@ GenTree* DecomposeLongs::DecomposeArith(LIR::Use& use)
 }
 
 //------------------------------------------------------------------------
+// DecomposeArith: Decompose GT_ADD, GT_SUB, GT_OR, GT_XOR, GT_AND.
+//
+// Arguments:
+//    use - the LIR::Use object for the def that needs to be decomposed.
+//
+// Return Value:
+//    The next node to process.
+//
+GenTree* DecomposeLongs::DecomposeShift(LIR::Use& use)
+{
+    assert(use.IsInitialized());
+
+    GenTree* tree   = use.Def();
+    GenTree* gtLong = tree->gtGetOp1();
+    genTreeOps oper = tree->OperGet();
+
+    assert((oper == GT_LSH) || (oper == GT_RSH) || (oper == GT_RSZ));
+
+    unsigned blockWeight = m_block->getBBWeight(m_compiler);
+
+    LIR::Use loOp1Use(BlockRange(), &gtLong->gtOp.gtOp1, gtLong);
+    loOp1Use.ReplaceWithLclVar(m_compiler, blockWeight);
+
+    LIR::Use hiOp1Use(BlockRange(), &gtLong->gtOp.gtOp2, gtLong);
+    hiOp1Use.ReplaceWithLclVar(m_compiler, blockWeight);
+
+    LIR::Use shift(BlockRange(), &tree->gtOp.gtOp2, tree);
+    shift.ReplaceWithLclVar(m_compiler, blockWeight);
+
+    GenTree* loOp1 = gtLong->gtGetOp1();
+    GenTree* hiOp1 = gtLong->gtGetOp2();
+
+    GenTree* shiftOp = tree->gtGetOp2();
+
+    GenTreePtr newLoOp1 = new (m_compiler, GT_LCL_VAR)
+        GenTreeLclVar(GT_LCL_VAR, loOp1->TypeGet(), loOp1->AsLclVarCommon()->GetLclNum(), BAD_IL_OFFSET);
```

Comment:
It leads to an assert in lir:
assert(slowNode->gtPrev == prevSlowNode);

I'm not exactly sure what this assert was saying, but moving them into new GT_LCL_VARs bypassed that issue. @pgavlin what exactly is this assert telling us?